python3Packages.lintrunner: init at 0.12.7

PrestonHager opened the pull request:
Based on https://github.com/suo/lintrunner upstream. Package is listed in PyPI already at https://pypi.org/project/lintrunner/.


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

natsukium commented on maintainers/maintainer-list.nix:
```diff
@@ -18467,6 +18467,13 @@
     githubId = 246631;
     keys = [ { fingerprint = "3E46 7EF1 54AA A1D0 C7DF  A694 E45C B17F 1940 CA52"; } ];
   };
+  prestonhager = {
```
Could you add yourself in another commit?

PrestonHager replied:
Added myself in the first commit, considering this resolved.

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+  python,
+}:
+
+buildPythonPackage rec {
```
I think this tool is an application but not a library.
If so, please use `buildPythonApplication` and move it to `pkgs/by-name/li/lintrunner` instead.

PrestonHager replied:
It is an application, I've used `buildPythonApplication` instead of `buildPythonPackage`. The typical way to install the application is via `pip`, should the package then be referenced/placed in the python top-level packages? This way it can be installed using `python3Packages.lintrunner` instead of the package name `lintrunner`, or is this bad practice?

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+buildPythonPackage rec {
+  pname = "lintrunner";
+  version = "0.12.7";
+  format = "pyproject";
```
```suggestion
  pyproject = true;
```

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+  # repository instead of a fork.
+  src = fetchFromGitHub {
+    owner = "PrestonHager";
+    repo = pname;
```
```suggestion
    repo = "lintrunner;
```
`repo` and `pname` are semantically different.

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+  src = fetchFromGitHub {
+    owner = "PrestonHager";
+    repo = pname;
+    rev = "main";
+    #rev = "v${version}";
```
```suggestion
    tag = "v${version}";
```
and please use `fetchpatch2`.

PrestonHager replied:
Unsure what you mean by `fetchpatch2`? I'd like to use patches instead of a fork of the repository, but I could not get a `git diff --cached > file.patch` to work locally, probably doing something wrong.

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+  };
+
+  # Adding maturninBuildHook will automatically build and install the executable
+  nativeBuildInputs = [
+    maturin
+    rustPlatform.cargoSetupHook
+    rustPlatform.maturinBuildHook
```
```suggestion
  build-system = [ rustPlatform.maturinBuildHook ];

  nativeBuildInputs = [
    rustPlatform.cargoSetupHook
```

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+    cargo
+    rustc
+  ];
+
```
Could you add tests with `pythonImportsCheck` and `pytestCheckHook`?
https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#testing-python-packages-testing-python-packages

PrestonHager replied:
Since it's an application, there's no imports to check with python. It's written in rust, should I run `cargo test` under the package to ensure tests? Not quite the same as testing the installed binary, but testing the application code functionality itself.

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+    rustc
+  ];
+
+  meta = with lib; {
```
```suggestion
  meta = {
```

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+    homepage = "https://github.com/suo/lintrunner";
+    changelog = "https://github.com/suo/lintrunner/releases/tag/v${version}";
+    mainProgram = "lintrunner";
+    license = licenses.mit;
```
dual license?
https://github.com/suo/lintrunner/blob/1b70da01a6a9237c0f60b07d017f6fb60f49ac8c/Cargo.toml#L7

PrestonHager replied:
I left an issue in the upstream repository, [#84](https://github.com/suo/lintrunner/issues/84). If I were to guess, it's meant to be MIT and the other one is from Cargo doing something. In the meantime, is there a dual license option we can do to specify both licenses for the nix package?

PrestonHager replied:
Didn't realize that license could be a list. So added the second license to the package definition.

natsukium commented on pkgs/development/python-modules/lintrunner/default.nix:
```diff
@@ -0,0 +1,55 @@
+  ];
+
+  meta = with lib; {
+    description = "A tool that runs linters.";
```
```suggestion
    description = "Tool that runs linters";
```
https://nixos.org/manual/nixpkgs/unstable/#var-meta-description

PrestonHager committed:
Added maintainer prestonhager

Added myself as a maintainer for future packages.

PrestonHager committed:
lintrunner: init at 0.12.7

Based on https://github.com/suo/lintrunner upstream. Package is listed
in PyPI already at https://pypi.org/project/lintrunner/.
Built using buildPythonApplication with maturin (rust) build system.

PrestonHager commented:
@natsukium I was able to make most of the changes you suggested, I have a couple of questions with regards to some changes. I left comments under each suggestion, but here they are below for completeness:

1. It is an application, I've used `buildPythonApplication` instead of `buildPythonPackage`. The typical way to install the application is via pip, should the package then be referenced/placed in the python top-level packages? This way it can be installed using `python3Packages.lintrunner` instead of the package name `lintrunner`, or is this bad practice?
2. I'd like to use patches instead of a fork of the repository, but I could not get a `git diff --cached > file.patch` to work locally, probably doing something wrong. I tried using `patches = [ file.patch ];` and `cargoPatches = [ file.patch ];` but neither seemed to change the contents of the source directory. I also noticed that this would be creating a new file (Cargo.lock) rather than changing the contents of a file.
3. Since it's an application, there's no imports to check with python. It's written in rust, should I run `cargo test` under the package to ensure tests? Not quite the same as testing the installed binary, but testing the application code functionality itself.
4. I left an issue in the upstream repository, [#84](https://github.com/suo/lintrunner/issues/84). If I were to guess, it's meant to be MIT and the other one is from Cargo doing something. In the meantime, is there a dual license option we can do to specify both licenses for the nix package?

Thank you for the review and all the help!

PrestonHager committed:
lintrunner: Removed unnecessary input

FliegendeWurst commented on pkgs/by-name/li/lintrunner/package.nix:
```diff
@@ -0,0 +1,52 @@
+    hash = "sha256-CYSNEDT9RO7bnVW8zcDDJOva4GOKGMNs1CcU//IbBrc=";
+  };
+
+  cargoDeps = rustPlatform.fetchCargoTarball {
```
Please switch to the newer fetchCargoVendor #378288

sikmir commented on pkgs/by-name/li/lintrunner/package.nix:
```diff
@@ -0,0 +1,52 @@
+  version = "0.12.7";
+  pyproject = true;
+
+  # NOTE: there is a PR open on the upstream repository for the Cargo.lock file.
```
Maybe fetchpatch is a better option here.

PrestonHager committed:
Merge remote-tracking branch 'origin/master' into add-lintrunner-package

PrestonHager committed:
lintrunner: Updated source to upstream repository

PrestonHager committed:
lintrunner: Switch to fetchCargoVendor

PrestonHager committed:
lintrunner: Use hash to pin the working version

PrestonHager committed:
lintrunner: Added BSD3 license to package defintion
